<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<title>Untitled Document</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<link href="PlasmID%20October%202005/PlasmID%20October%20Revision%20Finals/plasmidstyle.css" rel="stylesheet" type="text/css">
<link href="PlasmID%20October%202005/PlasmID%20October%20Revision%20Finals/plasmidstyle.css" rel="stylesheet" type="text/css">
</head>

<body>
<p class="mainbodytextlarge">Information about Cloning Methods</p>
<p class="featuretext">Included on this page:</p>
<p class="featuretext">Restriction Digest Approaches, Standard RE Digests<br>
  Restriction Digest Apporaches, CpoI-based Directional Cloning<br>
  Recombinational Approaches, Cre/Lox (P1 phage) approach (e.g. Clontech Creator 
  System)<br>
  Recombinational Approaches, att lambda phage approach (e.g. Invitrogen Gateway 
  System)<br>
  Bacterial Mating Approaches, MAGIC approach (Elledge Lab, HMS)</p>
<p class="mainbodytextlarge">Restriction Digest Approaches, Standard RE Digests</p>
<p class="featuretext">DESCRIPTION. In standard restriction enzyme (RE)cloning, 
  specific enzymes are used to cleave DNA at specific DNA sequences, and fragments 
  with compatable ends (or blunt-ended fragments) are ligated to one another to 
  form a circular plasmid, which can be transformed and propagated in bacteria. 
  The method can be used to add, remove or otherwise manipulate DNA sequences.</p>
<p class="featuretext">LEARN MORE. To learn more about this method, please consult 
  a molecular biology textbook or protocol manual.</p>
<p class="featuretext">ADVANTAGES. A significant advantage of RE cloning is that 
  because it has been in use for many years, reagents and protocols are readily 
  available to most researchers.</p>
<p class="featuretext"> IN THE DF/HCC DNA Resource Core Repository (PlasmID): 
  virtually any plasmid can be manipulated using restriction enzyme cloning methods. 
  Particularly useful are those plasmids that contain a multiple cloning site 
  (MCS), also called a polylinker. </p>
<p class="mainbodytextlarge">Restriction Digest Approaches, CpoI-based Directional 
  Cloning </p>
<p class="featuretext">DESCRIPTION. The CpoI enzyme (available from Fermentas 
  and other vendors) recognizes the sequence cggtccg or cggaccg. The enzyme cuts 
  after the first cg, leaving a three base-pair three-prime overhang. These properties 
  make it possible to use CpoI to clone specially designed PCR-amplified fragments 
  in-frame and directionally into a CpoI-digested vector. The CpoI approach can 
  be used to add inserts to a vector with aCpoI cloning site.</p>
<p class="featuretext">LEARN MORE. To learn more about this method, see papers 
  that use the method, such as Petrucco et al. (2002) &quot;A Nick-sensing ...&quot;, 
  Izymiya et al. (2003) &quot;Cell cycle regulation ...&quot; or Izumiya et al. 
  (2005) &quot;Kaposi's sarcoma-associated ...&quot;. In addition, it is helpful 
  to draw out the vector and PCR product ends to see how the method can work.</p>
<p class="featuretext">ADVANTAGES. Significant advantages of the CpoI-based directional 
  cloning method over conventional restriction digest methods include (1) as the 
  overhang is three base-pairs, it's easy to keep things in-frame, and (2) a single 
  enzyme can be used in a directional approach (most directional approaches require 
  that one end is cut with one restriction enzyme and the other end is cut with 
  a different restriction enzyme).</p>
<p class="featuretext">IN THE DF/HCC DNA Resource Core Repository (PlasmID): Several 
  plasmids shared with the repository by J. Kamil (Laboratory of D. Coen, BCMP 
  Department at Harvard Medical School) that are useful for adding N-terminal 
  and C-terminal tags (e.g. 6xHis, T7 eptiope tags). Try &quot;advanced search&quot; 
  with &quot;Kamil&quot; as author or try &quot;search by vector&quot; and choose 
  &quot;CpoI-based cloning&quot; as the cloning method.</p>
<p class="mainbodytextlarge">Recombinational Approaches, Cre/Lox (P1 phage) approach 
  (e.g. Clontech Creator System)</p>
<p class="featuretext">DESCRIPTION. Clontech's Creator (TM) system and other LoxP 
  site-containing plasmids can be used to add, replace or otherwise manipulate 
  DNA using the Cre enzyme, which catalyzes site-specific recombination at LoxP 
  sites. Please note that not every vector containing LoxP is appropriate for 
  Cre/Lox-based cloning approaches. Sometimes the LoxP is present to make possible 
  some experimental approach that is unrelated to cloning (for example, LoxP-containing 
  constructs are sometime introduced into cells and then in vivo expression of 
  Cre is used to bring about changes).</p>
<p class="featuretext">LEARN MORE. To learn more about this method, please see 
  the Clontech website www.clontech.com and appropriate references, which include 
  Hoess et al. (1982) P1 site-specific recombination: nucleotide sequence of the 
  recombining sites. PubMed ID 6954485.</p>
<p class="featuretext">ADVANTAGES. A significant advantage of the system is that 
  an insert can be cloned into a single 'master' or 'entry' vector and readily 
  sub-cloned into many different expression vectors without the need to digest 
  with restriction enzymes, gel purify, ligate, etc.</p>
<p class="featuretext">IN THE DF/HCC DNA Resource Core Repository (PlasmID): The 
  very many sequence-verified open reading frame (ORF) or cDNA clones in pDNR-Dual 
  shared with the repository by the Harvard Institute of Proteomics (HIP) at Harvard 
  Medical School. Some of these clones are &quot;closed&quot; format such that 
  the normal stop codon is present and others are &quot;fusion&quot; format in 
  which the stop codon is replaced by a leu codon so that a C-terminal tag can 
  be added.</p>
<p class="mainbodytextlarge">Recombinational Approaches, att lambda phage approach 
  (e.g. Invitrogen Gateway System)</p>
<p class="featuretext">DESCRIPTION. Invitrogen's Gateway (TM) system and other 
  lambda phage (att) recombination site-containing plasmids can be used to add, 
  replace or otherwise manipulate DNA using an enzyme that catalyzes site-specific 
  recombination at att-type sites.</p>
<p class="featuretext">LEARN MORE. To learn more about this method, please see 
  the Invitrogen website www.invitrogen.com and appropriate references, which 
  include Landy (1989) Dynamic, structural and regulatory aspects of lambda site-specific 
  recombination. PubMed ID 2528323.</p>
<p class="featuretext">ADVANTAGES. A significant advantage of the system is that 
  an insert can be cloned into a single 'master' or 'entry' vector and readily 
  sub-cloned into many different expression vectors without the need to digest 
  with restriction enzymes, gel purify, ligate, etc.</p>
<p class="featuretext">IN THE DF/HCC DNA Resource Core Repository (PlasmID): Examples 
  include the very many sequene-verified open reading frame (ORF) or cDNA clones 
  in pDONR201 or pDONR221 shared with the repository by the Harvard Institute 
  of Proteomics (HIP) at Harvard Medical School. Some of these clones are &quot;closed&quot; 
  format such that the normal stop codon is present and others are &quot;fusion&quot; 
  format in which the stop codon is replaced by a leu codon so that a C-terminal 
  tag can be added.</p>
<p class="mainbodytextlarge"> Bacterial Mating Approaches, MAGIC approach (Elledge 
  Lab, HMS)</p>
<p class="featuretext">DESCRIPTION. &quot;Mating-Assisted Genetically Integrated 
  Cloning&quot; or MAGIC was developed by Li &amp; Elledge (HHMI/Harvard Medical 
  School) and relies on in vivo DNA cleavage and homologous recombination events 
  that occur during bacterial mating to combine specifically designed 'donor' 
  and 'recipient' plasmids (e.g. insert-containing donors and specific recipient 
  expression vectors).</p>
<p class="featuretext">LEARN MORE. To learn more about this method, please see 
  Li &amp; Elledge (2005) MAGIC, an in vivo genetic method for the rapid construction 
  of recombinant DNA molecules. PubMed ID 15731760.</p>
<p class="featuretext">ADVANTAGES. A significant advantage of the method is that 
  the bacterial do the work of combining inserts and vectors and thus, constructs 
  can be generated without the need for restriction enzyme digest, gel purification, 
  etc. or use of recombination-promoting enzymes.</p>
<p class="featuretext">IN THE DF/HCC DNA Resource Core Repository (PlasmID): Examples 
  include the set of pPRIME-(marker)-recipient clones for shRNA-based approaches 
  that was shared with the repository by the Elledge lab (HHMI/Harvard Medical 
  School).</p>
</body>
</html>
