
<html>

<head>

</head>

<body background='background.gif'>
<h2>Help</h2>
<hr>
<table width="90%" border="1" cellspacing="2" cellpadding="2" align="center">
  <tr> 
    <td><div align="center"><strong>Process</strong></div></td>
    <td><div align="center"><strong>Description</strong></div></td>
  </tr>
  <tr> 
    <td>Run sequencing for end reads </td>
    <td><p><strong>User:</strong> selects what reads to run (forward/reverse/both 
        )</p>
      <p><strong>System:</strong></p>
      <ul>
        <li> Creates new plates for end reads (one plate copy per read type)</li>
        <li>Creates naming file and sends it to user by e-mail</li>
      </ul></td>
  </tr>
  <tr> 
    <td width="40%"> Run end read wrapper </td>
    <td width="60%">Fore execution of end reads wrapper (file transfer, Phred 
      base call, reads information parsing)</td>
  </tr>
  <tr> 
    <td>Check whether all reads are available</td>
    <td>System checks that all required runs to be complete (for plates that requires 
      both reads to be run only)</td>
  </tr>
  <tr> 
    <td><table width="100%" border="0" cellspacing="2" cellpadding="2">
        <tr> 
          <td> 1. Run isolate ranker</td>
        </tr>
        <tr> 
          <td> 2. Run polymorphism finder for isolate ranking </td>
        </tr>
      </table></td>
    <td><p><strong>User: </strong></p>
      <ul>
        <li>select settings for isolate ranker </li>
        <li>select plates(reads originated from the same glycerol plate selected 
          as one entity)</li>
      </ul>
      <p><strong>System:</strong></p>
      <ul>
        <li>executes isolate ranker</li>
        <li>executes polymorphism finder (optional, but can not be run separately)</li>
      </ul></td>
  </tr>
  <tr>
    <td>View, change (optional) and approve isolate ranking</td>
    <td><strong>User:</strong> selects plate name
      <p><strong>System:</strong> present output of isolate ranker to the user</p>
      </td>
  </tr>
  <tr colspan=2> 
    <td colspan="2"><table width="100%" border="3" cellpadding="2" cellspacing="2">
        <tr> 
          <td width="40%"> Send sequencing plates to the outside facility</td>
          <td width="60%"><p><strong>System:</strong> </p>
            <ul>
              <li>Create sequencing plates (can rearray plates)</li>
              <li>creates files with sequence information to be send with plate 
                to the sequencing facility</li>
            </ul></td>
        </tr>
        <tr> 
          <td> Receive sequencing plates from the outside facility</td>
          <td><p><strong>User: </strong>notify user what plates came back</p>
            <p><strong>System: </strong>changes plate location</p></td>
        </tr>
        <tr> 
          <td> Upload data from outside facility</td>
          <td><strong>System</strong>: upload assembled sequences from file. If 
            uploaded data are for plates that has been not analyzed by ACE for 
            end reads, sequencing plates are created.</td>
        </tr>
      </table></td>
    
  </tr>
  <tr> 
    <td colspan="2"> <table width="100%" border="3" cellspacing="2" cellpadding="2">
        <tr> 
          <td width="40%"> Run primer designer</td>
          <td width="60%"><p><strong>User:</strong> select plates for full sequencing 
              (system should know what sequences from these plates were covered 
              by end reads)</p>
            <p><strong>System:</strong></p>
            <ul>
              <li>Design primers</li>
              <li>Create template plates</li>
              <li>Create oligo plates</li>
              <li>Create file for oligo order</li>
            </ul></td>
        </tr>
        <tr> 
          <td height="26">Receive oligo plates</td>
          <td><p><strong>User: </strong>notify system what plates came back</p>
            <p><strong>System: </strong>changes plate location</p></td>
        </tr>
        <tr>
          <td>Run full sequencing process</td>
          <td><strong>User: </strong>select template plates 
            <p><strong>System: </strong>creates naming files</p></td>
        </tr>
        <tr> 
          <td> Run assembly wrapper </td>
          <td>Forse execution of assembly wrapper (file transfer, assembly, sequence 
            information parsing)</td>
        </tr>
      </table></td>
 
  </tr>
  <tr> 
    <td>Run discrepancy finder for clone evaluation </td>
    <td><p><strong>User: </strong></p>
      <ul>
        <li>select settings</li>
        <li>select template plates </li>
      </ul>
      <p><strong>System: </strong>run discrepancy finder for isolates on selected 
        plates </p></td>
  </tr>
  <tr> 
    <td>Run polymorphism finder for clone evaluation </td>
    <td><p><strong>User: </strong></p>
      <ul>
        <li>select settings</li>
        <li>select template plates </li>
      </ul>
      <p><strong>System: </strong>run polymorphism finder for isolates on selected 
        plates </p></td>
  </tr>
  <tr> 
    <td>Run decision tool </td>
    <td><p><strong>User: </strong></p>
      <ul>
        <li>select settings</li>
        <li>select template plates </li>
      </ul>
      <p><strong>System: </strong>run decision tool for isolates on selected plates 
        and provide results to user</p>
      <p>User: approves results with changes (optional) </p>
      <p>System: send rearray file to user</p></td>
  </tr>
  <tr> 
    <td>Storage plate assembly </td>
    <td>&nbsp;</td>
  </tr>
</table>
<p>
<p> 

</body>
</html>


